### ISSUE TITLE
Profiling

### ISSUE THREAD

**Issue by LoganDark** [fr1kin/ForgeHax#107]
I want to see what's sucking up all my frames, but here's what I see:

![image](https://user-images.githubusercontent.com/4723091/63310674-75ecda00-c2b0-11e9-84b1-97324c6156ec.png)

Unspecified!!

**Comment by LoganDark** [comment]
*turns off markers, percent drops to 0.1%*

**Comment by fr1kin** [comment]
yes markers is really hacky and needs to be completely redone

**Comment by LoganDark** [comment]
Since ForgeHax seems to be pretty far from Forge's shitty APIs I might consider trying to rewrite, would you be willing to motivate me?? lol

**Comment by fr1kin** [comment]
Yes like I said markers is way too complicated. It works by injecting code in minecrafts chunk render workers and uses them to build the VBOs. It was the lazy solution and I thought it would be just fine, but it turns out it just caused more complications. The proper way to do it would be to do what schematica does (essentially copy all of minecrafts render worker code), but of course that would have required I wrote a ton of code. There is another solution that my friend uses and it seems to work very well with minimal code, so I'll see if I can just copy that.

**Comment by LoganDark** [comment]
Oooooorrrrrr you can just keep a cache of tracked blocks and run some OpenGL code at the end of every frame.

**Comment by fr1kin** [comment]
There is. You can't just have one giant VBO that all loaded blocks fall under because new chunks generate, old ones get removed, and chunks are updated. Scanning every loaded block is very costly, so it must be done across multiple threads.

**Comment by LoganDark** [comment]
> You can't just have one giant VBO that all loaded blocks fall under

I'm confused by what you mean. Of course there will be a cost if you add stone or bedrock. It can't get much worse than what's going on here.

> new chunks generate, old ones get removed, and chunks are updated

You can update the cache from another thread as this happens. New chunk? Scan all the blocks. Block update? Update the cache accordingly. Chunk removed? Remove that chunk from the cache.

Adding a new block would re-scan all the chunks, of course, but that doesn't happen often so it wouldn't be too much unlike F3+A, which is quite fast anyway.

> Scanning every loaded block is very costly, so it must be done across multiple threads.

I sincerely hope you are not doing this every frame.

**Comment by LoganDark** [comment]
Should I rename/rewrite this issue to be about the Markers mod specifically, and move the profiling thing out into a separate issue?

**Comment by fr1kin** [comment]
>I'm confused by what you mean. Of course there will be a cost if you add, for example, stone or bedrock. It can't get much worse than what's going on here.
A VBO is a buffer object that is uploaded to the GPU once, and then rendered in the main opengl thread. This is done to prevent having to recalculate the buffer every frame and upload it (this would be very costly). These buffers can vary in size, but are generally small so as to allow them to be uploaded very quickly. Each buffer is a RenderChunk object, which is a 16x16x16 area. Once uploaded, VBOs generally cant be updated without being reuploaded again.

>You can update the cache from another thread as this happens. New chunk? Scan all the blocks. Block update? Update the cache accordingly. Chunk removed? Remove that chunk from the cache.

I wish it were this simple, but it isnt. Keep in mind there are multiple workers running concurrently. Concurrency isn't simple, and generally you don't want any of the workers running to access shared memory. That would require locks, which would require the worker to pause until it can is safe to access said shared memory. All that can cause deadlocks and even more problems. Concurrency and multithreading can get very complicated very quickly, and what mojang has written works perfectly with no issues.

>Removing a block can just scan the cache and remove it. It can get quite efficient.

I am not going to go too deep into details, but it's far from that simple.

>I sincerely hope you are not scanning every block every frame.

If I did that the game would run at sub 1fps.

**Comment by fr1kin** [comment]
>Should I rename/rewrite this issue to be about the Markers mod specifically, and move the profiling thing out into a separate issue?

Sure

**Comment by LoganDark** [comment]
> A VBO is a buffer object that is uploaded to the GPU once, and then rendered in the main opengl thread. This is done to prevent having to recalculate the buffer every frame and upload it (this would be very costly). These buffers can vary in size, but are generally small so as to allow them to be uploaded very quickly. Each buffer is a RenderChunk object, which is a 16x16x16 area. Once uploaded, VBOs generally cant be updated without being reuploaded again.

I'm curious as to what the performance hit would be if you did recalculate it every frame. The buffer, assuming all marked blocks are correctly cached.

> I wish it were this simple, but it isnt. Keep in mind there are multiple workers running concurrently. Concurrency isn't simple, and generally you don't want any of the workers running to access shared memory. That would require locks, which would require the worker to pause until it can is safe to access said shared memory. All that can cause deadlocks and even more problems. Concurrency and multithreading can get very complicated very quickly, and what mojang has written works perfectly with no issues.

As long as operations to the cache are atomic, I think that won't matter? After all, if blocks are being updated while a frame is being rendered, the worst that could happen is that the iteration gets kind of screwed up and you get a glitch for a few frames, assuming you're not using fail-fast...

> I am not going to go too deep into details, but it's far from that simple.

How is it "far from that simple"? I'm gonna have to ask for more details here, because an assertion like that should come with an explanation... I don't understand how you can't remove a block from the cache if it gets removed (like `markers options remove mossy_cobblestone`).

**Comment by LoganDark** [comment]
LOL what is that label?

![image](https://user-images.githubusercontent.com/4723091/63331672-03e5b680-c2eb-11e9-904b-111f041d2500.png)

Looking at the description of the label and other issues marked with it I wouldn't consider this a "dumb bitch" issue. I may not be the most educated but I'm not exactly a "dumb bitch". :/

**Comment by VADemon** [comment]
On a side note some other mods that modify the renderer (old school Xray in my case: Xray-DLB) break the markers mod. (Didn't investigate)

**Comment by LoganDark** [comment]
@VADemon This issue is about Markers' speed, not compatibility. A rewrite should fix that however, you should make a separate issue (about compatibility, not a rewrite specifically).

**Comment by VADemon** [comment]
@logandark Kindly thank you, I have not noticed yet, Sir.

**Comment by wifebeater** [comment]
works on my pc

**Comment by LoganDark** [comment]
@wifebeater excuse me?

**Comment by fr1kin** [comment]
@0-x-2-2 don't give that label to people that don't deserve it.
